feat(swarm): allow behaviours to share addresses of peers

StemCll opened the pull request:
## Description

Resolves: #4302.

<!--
Please write a summary of your changes and why you made them.
This section will appear as the commit message after merging.
Please craft it accordingly.
For a quick primer on good commit mesages, check out this blog post: https://cbea.ms/git-commit/

Please include any relevant issues in here, for example:

Related https://github.com/libp2p/rust-libp2p/issues/ABCD.
Fixes https://github.com/libp2p/rust-libp2p/issues/XYZ.
-->

## Notes & open questions
Still WIP

<!--
Any notes, remarks or open questions you have to make about the PR which don't need to go into the final commit message.
-->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2551,6 +2551,7 @@ where
             }
             FromSwarm::DialFailure(dial_failure) => self.on_dial_failure(dial_failure),
             FromSwarm::AddressChange(address_change) => self.on_address_change(address_change),
+            FromSwarm::NewExternalAddrOfPeer(_addr) => todo!(),
```
This is the place where the new event comes in handy. Here we can add the new address to our routing table.

mxinden commented on swarm/src/lib.rs:
```diff
@@ -1112,6 +1113,37 @@ where
+                        .values()
+                        .flatten()
+                        .filter_map(|server| self.transport.address_translation(server, &addr))
+                        .collect();
```
Given that this is the address of a remote and not a local address, I don't think there is any value in `address_translation`. Likely just a copy and paste from the `NewAddrCandidate` event.

StemCll replied:
:heavy_check_mark: 

Left just `on_swarm_event` call

mxinden replied:
Sounds good. Let me know once you need another review here.

mxinden reviewed (commented):
This is going in the right direction. Thank you for the work!

StemCll commented on protocols/kad/src/behaviour/test.rs:
```diff
@@ -1378,6 +1378,89 @@ fn network_behaviour_on_address_change() {
+    assert!(matches!(
+        kademlia.kbuckets.entry(&key),
+        kbucket::Entry::Present(_, _)
+    ));
```
is it worth testing here?

thomaseizinger commented on protocols/autonat/src/behaviour.rs:
```diff
@@ -589,6 +589,12 @@ impl NetworkBehaviour for Behaviour {
+                peer_id: _peer_id,
+            }) => {
+                todo!();
+            }
```
I don't think autonat needs to know about new external addresses of a peer :)

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on protocols/identify/src/behaviour.rs:
```diff
@@ -395,6 +395,7 @@ impl NetworkBehaviour for Behaviour {
                     }
                 }
             }
+            FromSwarm::NewExternalAddrOfPeer(_addr) => todo!(),
```
FIXME: Remove this `todo!`.

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on protocols/request-response/src/lib.rs:
```diff
@@ -775,15 +775,16 @@ where
-            FromSwarm::ListenerClosed(_) => {}
-            FromSwarm::NewExternalAddrCandidate(_) => {}
-            FromSwarm::ExternalAddrExpired(_) => {}
-            FromSwarm::ExternalAddrConfirmed(_) => {}
```
I'd be great to avoid this kind of diff. Happy to review a PR that streamlines this across the repository though!

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on swarm-derive/src/lib.rs:
This change needs a minor version bump of `libp2p-swarm-derive` because it is not backwards-compatible with previous versions!

thomaseizinger commented on swarm/src/behaviour.rs:
```diff
@@ -320,6 +320,10 @@ pub enum ToSwarm<TOutEvent, TInEvent> {
+
+    /// TODO: more documentation needed
+    /// Reports external address of a remote peer to other `NetworkBehaviours`
+    NewExternalAddrOfPeer { addr: Multiaddr, peer_id: PeerId },
```
Intuitively, I'd have expected these to be flipped, i.e. for the `PeerId` to be at the front.

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on swarm/src/behaviour.rs:
```diff
@@ -561,6 +573,13 @@ pub struct ExternalAddrExpired<'a> {
+#[derive(Clone, Copy, Debug)]
+pub struct NewExternalAddrOfPeer<'a> {
+    pub addr: &'a Multiaddr,
+    pub peer_id: &'a PeerId,
```
Same here, can you flip these?

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on swarm/src/lib.rs:
```diff
@@ -1580,9 +1588,9 @@ impl fmt::Display for DialError {
             ),
-            DialError::DialPeerConditionFalse(PeerCondition::Disconnected) => write!(f, "Dial error: dial condition was configured to only happen when disconnected (`PeerCondition::Disconnected`), but node is already connected, thus cancelling new dial."),
-            DialError::DialPeerConditionFalse(PeerCondition::NotDialing) => write!(f, "Dial error: dial condition was configured to only happen if there is currently no ongoing dialing attempt (`PeerCondition::NotDialing`), but a dial is in progress, thus cancelling new dial."),
-            DialError::DialPeerConditionFalse(PeerCondition::Always) => unreachable!("Dial peer condition is by definition true."),
```
Can you undo this diff? Looks like a bad merge!

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2551,6 +2583,7 @@ where
             }
             FromSwarm::DialFailure(dial_failure) => self.on_dial_failure(dial_failure),
             FromSwarm::AddressChange(address_change) => self.on_address_change(address_change),
+            FromSwarm::NewExternalAddrOfPeer(addr) => self.on_new_address_event(addr),
```
Wouldn't it be simpler to just delegating to `Kademlia::add_address` here?

StemCll replied:
yep :)

thomaseizinger commented on protocols/kad/src/behaviour/test.rs:
```diff
@@ -1378,6 +1378,89 @@ fn network_behaviour_on_address_change() {
 }
 
+#[test]
+fn network_behaviour_on_new_external_address_of_peer_for_present_entry() {
```
If we delegate the event to `Kademlia::add_address`, I don't think we need a dedicated test for it!

thomaseizinger reviewed (commented):
Great progress! I've left a few comments :)

What I'd also like to see in this PR is:

- mDNS emitting `ToSwarm::NewExternalAddrOfPeer`
- identify emitting `ToSwarm::NewExternalAddrOfPeer` for the self-advertised listen addresses of a peer
- the rendezvous client emitting `ToSwarm::NewExternalAddrOfPeer` whenever it receives a peer record

Something that we should discuss is whether we want to add a public function to `Swarm` that allows users to trigger this event without having to call particular behaviour. For example, at the moment, `libp2p-request-response` offers a function `add_address` to populate an internal map of peer addresses. I'd like that function to go away because I think it is none of `libp2p-request-response`s business to do address management.

If we were to add a function to `Swarm` that triggers this event, we could deprecate the function in `libp2p-request-response` and instead add an internal cache that stores peer addresses. To avoid duplicated code across multiple behaviours, we could add a `PeerAddresses` to `libp2p-swarm` similar to how we have [`ExternalAddresses`](https://github.com/libp2p/rust-libp2p/blob/master/swarm/src/behaviour/external_addresses.rs).

I don't think all of this has to be done in one PR but what we do should be consistent. For example, I think it would be to merge this whilst also deprecating `libp2p-request-response`'s `add_address` function to point to the new `Swarm` function instead. We can always introduce the `PeerAddresses` struct at a later point because it is a backwards-compatible change.

thomaseizinger commented on swarm/src/behaviour.rs:
```diff
@@ -320,6 +320,10 @@ pub enum ToSwarm<TOutEvent, TInEvent> {
     },
+
+    /// TODO: more documentation needed
+    /// Reports external address of a remote peer to other `NetworkBehaviours`
```
```suggestion
    /// Reports external address of a remote peer to other [`NetworkBehaviours`].
```

thomaseizinger commented on swarm-derive/CHANGELOG.md:
```diff
@@ -1,4 +1,11 @@
-## 0.33.0 
+## 0.34.0
```
```suggestion
## 0.34.0 - unreleased
```

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on swarm-derive/CHANGELOG.md:
```diff
@@ -1,4 +1,11 @@
-## 0.33.0 
+## 0.34.0
+
+- Add `from_swarm::NewExternalAddrOfPeer` event.
```
I think this should say something along the lines of: "Generate code for ..."

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on swarm/src/behaviour.rs:
```diff
@@ -320,6 +320,10 @@ pub enum ToSwarm<TOutEvent, TInEvent> {
+
+    /// TODO: more documentation needed
+    /// Reports external address of a remote peer to other `NetworkBehaviours`
+    NewExternalAddrOfPeer { peer_id: PeerId, addr: Multiaddr },
```
Adding this is also a breaking change for exhaustive matches of this enum so we should also bump the minor version of this crate and add a changelog entry.

StemCll replied:
:heavy_check_mark:

mxinden commented on swarm/src/behaviour.rs:
```diff
@@ -448,6 +458,8 @@ pub enum FromSwarm<'a, Handler> {
     ExternalAddrConfirmed(ExternalAddrConfirmed<'a>),
     /// Informs the behaviour that an external address of the local node expired, i.e. is no-longer confirmed.
     ExternalAddrExpired(ExternalAddrExpired<'a>),
+    /// TODO: document it
```
```suggestion
    /// Informs the behaviour that we have discovered a new external address for a remote peer.
```

StemCll replied:
:heavy_check_mark:

mxinden commented on swarm/src/behaviour.rs:
```diff
@@ -561,6 +573,13 @@ pub struct ExternalAddrExpired<'a> {
     pub addr: &'a Multiaddr,
 }
 
+/// [`FromSwarm`] variant that informs the behaviour that a new external address was detected.
```
```suggestion
/// [`FromSwarm`] variant that informs the behaviour that a new external address for a remote peer was detected.
```

StemCll replied:
:heavy_check_mark:

mxinden reviewed (commented):
As always, appreciate the thorough work @StemCll!

thomaseizinger commented on protocols/identify/src/behaviour.rs:
```diff
@@ -291,11 +291,15 @@ impl NetworkBehaviour for Behaviour {
+                    .push_back(ToSwarm::NewExternalAddrCandidate(observed_addr.clone()));
+                self.events.push_back(ToSwarm::NewExternalAddrOfPeer {
+                    peer_id,
+                    addr: observed_addr,
```
This is the address that the other node observed of _us_. What we want to emit here are the self-advertised addresses of the other peer contained in `info`!

thomaseizinger replied:
I'd like to see a test for this:

- Connect two swarms
- One swarm adds an external address
- 2nd swarm receives this via identify push and emits this event
- assert that 2nd swarm receives this event

StemCll replied:
> What we want to emit here are the self-advertised addresses of the other peer contained in info!

Hmm, so in that case should the `NewExternalAddrOfPeer` have Vec of `Multiaddr`s? Should we handle a case where we get multiple new external addresses in `Info`?

thomaseizinger replied:
> > What we want to emit here are the self-advertised addresses of the other peer contained in info!
> 
> Hmm, so in that case should the `NewExternalAddrOfPeer` have Vec of `Multiaddr`s? Should we handle a case where we get multiple new external addresses in `Info`?

I don't think there is a need for a `Vec`, we can just emit multiple instances of the same event right?

StemCll replied:
I've added related changes - struggling a bit with a smoke test for swarms atm

thomaseizinger replied:
Are you joining the [maintainers call](https://github.com/libp2p/rust-libp2p/discussions/4606) tomorrow? Happy to talk about it here.

StemCll replied:
Hi @thomaseizinger :wave: 
I'll join the next maintainers call to talk about these changes

thomaseizinger replied:
Awesome, looking forward to it!

thomaseizinger commented on swarm-derive/CHANGELOG.md:
```diff
@@ -1,4 +1,11 @@
-## 0.33.0 
+## 0.34.0 - unreleased
+
+- Generate code for `libp2p-swarm`'s `from_swarm::NewExternalAddrOfPeer` enum variant.
```
```suggestion
- Generate code for `libp2p-swarm`'s `From_swarm::NewExternalAddrOfPeer` enum variant.
```

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on swarm/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+## 0.44.0
```
```suggestion
## 0.44.0 - unreleased
```

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on protocols/identify/Cargo.toml:
```diff
@@ -3,7 +3,7 @@ name = "libp2p-identify"
 rust-version = { workspace = true }
 description = "Nodes identifcation protocol for libp2p"
-version = "0.43.0"
+version = "0.43.1"
```
Because `libp2p-swarm` is a dependency that is exposed in the public API, this is actually a breaking change for all protocol crates in one go.

For now, it is okay to leave this as is. We will make a PR that "open"s the merge window for breaking changes which bumps the versions accordingly.

thomaseizinger commented on swarm/src/behaviour.rs:
```diff
@@ -320,6 +320,9 @@ pub enum ToSwarm<TOutEvent, TInEvent> {
         connection: CloseConnection,
     },
+
+    /// Reports external address of a remote peer to other [`NetworkBehaviour`].
```
```suggestion
    /// Reports external address of a remote peer to the [`Swarm`] and through that to other [`NetworkBehaviour`]s.
```

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on swarm/src/behaviour.rs:
```diff
@@ -561,6 +572,13 @@ pub struct ExternalAddrExpired<'a> {
+/// [`FromSwarm`] variant that informs the behaviour that a new external address for a remote peer was detected.
+#[derive(Clone, Copy, Debug)]
+pub struct NewExternalAddrOfPeer<'a> {
+    pub peer_id: &'a PeerId,
```
PeerId is `Copy` so we don't need a reference to it.

StemCll replied:
:heavy_check_mark:

thomaseizinger reviewed (commented):
Thanks for addressing the continued work!

I made some more comments :)
Plus, I'd still like to see the `Swarm::add_address_of_peer` function and deprecate `Kademlia::add_address` with a note pointing to the other function. In the future, we will then make that private.

thomaseizinger commented on protocols/identify/src/behaviour.rs:
```diff
@@ -519,4 +549,38 @@ mod tests {
+        old_addrs.append(&mut new_addrs.clone());
+
+        assert_eq!(filter_new_external_addrs(&old_addrs, &new_addrs), None)
+    }
```
I think I'd rather not like to rely on randomness here. Hardcoded numbers should make this easier to reason about. Plus, you can then make a utility function for giving you a `/memory` address where you just pass a `u64` in. That should shorten these tests a lot.

Commit:
swarm: Add NewExternalAddrOfPeer

Commit:
deprecate request-response add_address

Commit:
swarm: add peer addresses

StemCll commented on swarm/src/behaviour/peer_addresses.rs:
```diff
@@ -0,0 +1,126 @@
+            }
+        }
+        false
+    }
```
@thomaseizinger @mxinden On which event should the address (or all peer's addresses) should be removed? Connection closed? :thinking:

StemCll replied:
Also - wouldn't it just be better to use `PeerCache` from identify?

thomaseizinger replied:
Yes, This definitely needs to be bounded in size using an LRU cache makes sense. I think we can pretty much move the `PeerCache` from identify to here and use this within `identify`.

thomaseizinger replied:
> @thomaseizinger @mxinden On which event should the address (or all peer's addresses) should be removed? Connection closed? 🤔

Part of what we want to achieve with this cache is that you can **re**-connect to a peer after the connection has been closed.

But we can remove it on `DialFailure` because clearly, the address doesn't work! Although we should check on _why_ the dial failed and e.g. don't remove it if it is just because it was `Denied`.

StemCll commented on protocols/autonat/src/behaviour.rs:
```diff
@@ -271,6 +271,7 @@ impl Behaviour {
     pub fn add_server(&mut self, peer: PeerId, address: Option<Multiaddr>) {
         self.servers.insert(peer);
         if let Some(addr) = address {
+            #[allow(deprecated)]
```
leaving this for now

thomaseizinger commented on swarm/src/lib.rs:
```diff
@@ -340,6 +341,9 @@ where
     listened_addrs: HashMap<ListenerId, SmallVec<[Multiaddr; 1]>>,
 
+    /// Cache of peers with their known external addresses.
+    peer_addrs: PeerAddresses,
```
I don't think the `Swarm` itself should keep a cache of addresses, I'd rather leave this to the individual modules.

thomaseizinger commented on swarm/src/lib.rs:
```diff
@@ -580,6 +585,11 @@ where
+    /// List external addresses of a known peer.
+    pub fn peer_addresses(&self, peer_id: &PeerId) -> Option<&Vec<Multiaddr>> {
+        self.peer_addrs.peer_addrs(peer_id)
+    }
```
Following from the above comment, I don't think we should/want to have this.

thomaseizinger commented on protocols/request-response/src/lib.rs:
```diff
@@ -470,6 +470,7 @@ where
     ///
     /// Returns true if the address was added, false otherwise (i.e. if the
     /// address is already in the list).
+    #[deprecated(note = "Use `libp2p_swarm::add_address` instead.")]
```
```suggestion
    #[deprecated(note = "Use `libp2p_swarm::Swarm::add_peer_address` instead.")]
```

thomaseizinger commented on protocols/request-response/src/lib.rs:
```diff
@@ -470,6 +470,7 @@ where
     /// address is already in the list).
+    #[deprecated(note = "Use `libp2p_swarm::add_address` instead.")]
     pub fn add_address(&mut self, peer: &PeerId, address: Multiaddr) -> bool {
         self.addresses.entry(*peer).or_default().insert(address)
```
`self.addresses` could now be the new `PeerCache`.

thomaseizinger replied:
Plus, we need to consume the `NewExternalAddrOfPeer` event and populate `self.addresses` from it, otherwise users are not going to get the same behaviour if they follow the deprecation advice and replace this call with `Swarm::add_peer_address`.

thomaseizinger commented on swarm/src/behaviour/peer_addresses.rs:
```diff
@@ -0,0 +1,126 @@
+
+    /// Returns peer's external addresses.
+    /// Returns [`None`] if the peer did not have any addrs cached.
+    pub fn peer_addrs(&self, peer_id: &PeerId) -> Option<&Vec<Multiaddr>> {
```
I think `impl Iterator` would be a better return value here.

thomaseizinger reviewed (commented):
Great progress! Thank you for reviving this PR :)

I'd like at least 1 behaviour to also _emit_ the new event. I think `libp2p-identify` makes the most sense here. Plus, it would be great if we could add a test for this. Ideally in `libp2p-request-response` as that is where we deprecate the function.

Commit:
Move peer cache to peer addresses

Commit:
Add pop function to peer address

thomaseizinger commented on protocols/request-response/src/lib.rs:
```diff
@@ -803,7 +797,13 @@ where
+                self.addresses.on_swarm_event(&event);
+            }
+            FromSwarm::NewExternalAddrOfPeer(_) => {
+                self.addresses.on_swarm_event(&event);
```
This seems unnecessarily redundant. I'd say the ergonomics of the API comes from taking `&FromSwarm` so that we _don't_ have to pattern match and "know" which events are interesting for this particular struct.

I'd suggest to just call `self.addresses.on_swarm_event` before this `match`.

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on swarm/src/behaviour/peer_addresses.rs:
```diff
@@ -0,0 +1,227 @@
+    fn prepare_addr(peer: &PeerId, addr: &Multiaddr) -> Multiaddr {
+        let addr = addr.clone();
+        addr.clone().with_p2p(*peer).unwrap_or(addr)
+    }
```
Let's make this a free function if it doesn't need `self`.

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on swarm/src/behaviour/peer_addresses.rs:
```diff
@@ -0,0 +1,227 @@
+
+    /// Removes address from peer addresses cache.
+    /// Returns true if the address was removed.
+    pub fn pop(&mut self, peer: &PeerId, address: &Multiaddr) -> bool {
```
I associate `pop` with a data structure that has has an order.

For one, `HashSet`s - which form the implementation detail of this data structure - don't have one.
Two, we shouldn't expose the implementation details of this.

I'd suggest calling it `remove`.

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on swarm/src/behaviour/peer_addresses.rs:
```diff
@@ -0,0 +1,227 @@
+
+    pub fn get_mut(&mut self, peer: &PeerId) -> Option<&mut HashSet<Multiaddr>> {
+        self.0.get_mut(peer)
+    }
```
Do we need this to be public? If not, I'd suggest limiting the public API as much as possible.

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on swarm/src/behaviour/peer_addresses.rs:
```diff
@@ -0,0 +1,227 @@
+                        _ => return false,
+                    }
+                }
+            }
```
Collapsing patterns reduces indentation which IMO is easier to read:

```suggestion
            FromSwarm::DialFailure(DialFailure { peer_id: Some(peer_id), error: DialError::NoAddresses, .. }) => {
                return self.0.pop(peer_id).is_some()
            }
```

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on swarm/src/behaviour/peer_addresses.rs:
```diff
@@ -0,0 +1,227 @@
+    }
+
+    /// Returns peer's external addresses.
+    pub fn get(&mut self, peer: &PeerId) -> Vec<Multiaddr> {
```
Let's return an iterator here so we don't eagerly allocate.

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on Cargo.toml:
```diff
@@ -99,8 +99,8 @@ libp2p-relay = { version = "0.17.1", path = "protocols/relay" }
-libp2p-swarm = { version = "0.44.0", path = "swarm" }
-libp2p-swarm-derive = { version = "=0.34.0", path = "swarm-derive" } # `libp2p-swarm-derive` may not be compatible with different `libp2p-swarm` non-breaking releases. E.g. `libp2p-swarm` might introduce a new enum variant `FromSwarm` (which is `#[non-exhaustive]`) in a non-breaking release. Older versions of `libp2p-swarm-derive` would not forward this enum variant within the `NetworkBehaviour`  ...
+libp2p-swarm = { version = "0.44.1", path = "swarm" }
+libp2p-swarm-derive = { version = "=0.35.0", path = "swarm-derive" }                            # `libp2p-swarm-derive` may not be compatible with different `libp2p-swarm` non-breaking releases. E.g. `libp2p-swarm` might introduce a new enum variant `FromSwarm` (which is `#[non-exhaustive]`) in a non-breaking release. Older versions of `libp2p-swarm-derive` would not forward this enum variant wit ...
```
Please avoid unrelated reformattings.

StemCll replied:
:heavy_check_mark:

thomaseizinger reviewed (commented):
Thank you for pushing this forward!

This looks pretty good now. CI won't run until the conflicts are resolved though.
I've left a few more but mostly small comments :)

I would imagine that the `request_response::Behaviour::add_address` function is actually used in a bunch of places that are now being flagged as using deprecated functions. We will need to change those.

Commit:
Add CR changes

Commit:
Remove public get_mut from peer addresses

Commit:
identify: emit  when identified new peer; add changelog entries and versions

thomaseizinger commented on protocols/identify/Cargo.toml:
```diff
@@ -3,7 +3,7 @@ name = "libp2p-identify"
 rust-version = { workspace = true }
 description = "Nodes identifcation protocol for libp2p"
-version = "0.44.1"
+version = "0.45.0"
```
We shouldn't need a minor version bump I think. Now that the enums are `non_exhaustive`, this shouldn't be a breaking change :)

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on protocols/request-response/Cargo.toml:
```diff
@@ -3,7 +3,7 @@ name = "libp2p-request-response"
 rust-version = { workspace = true }
 description = "Generic Request/Response Protocols"
-version = "0.26.1"
+version = "0.27.0"
```
Same here, this is backwards compatible so we can just bump this to `0.26.2`.

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on swarm-derive/Cargo.toml:
```diff
@@ -3,7 +3,7 @@ name = "libp2p-swarm-derive"
 rust-version = { workspace = true }
 description = "Procedural macros of libp2p-swarm"
-version = "0.34.1"
+version = "0.35.1"
```
```suggestion
version = "0.34.2"
```

This should work too I think. We pin the version so there shouldn't be a need for a minor version bump.

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on swarm/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+## 0.44.2 - unreleased
+
+- Add `FromSwarm::NewExternalAddrOfPeer` event.
+  See [PR 4371](https://github.com/libp2p/rust-libp2p/pull/4371)
```
Let's also document the new `Swarm` API please. That is what users will interact with most.

StemCll replied:
Added more info, but I'm not sure if it's sufficient for someone using the library :thinking:

thomaseizinger commented on swarm/src/behaviour/peer_addresses.rs:
```diff
@@ -0,0 +1,226 @@
+
+    pub fn new(cache_size: NonZeroUsize) -> Self {
+        Self(LruCache::new(cache_size))
+    }
```
Let's move this to the top please.

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on swarm/src/behaviour/peer_addresses.rs:
```diff
@@ -0,0 +1,226 @@
+        let addresses = addresses.filter_map(|a| a.with_p2p(peer).ok());
+        self.0.put(peer, HashSet::from_iter(addresses));
+        true
+    }
```
This seems like a strange API to me, why would we overwrite all addresses that are already in there? I think we should replace this with an `add` function that simply adds the address to the cache.

StemCll replied:
ach yes... my bad
:heavy_check_mark:

thomaseizinger commented on swarm/src/behaviour/peer_addresses.rs:
```diff
@@ -0,0 +1,226 @@
+            .cloned()
+            .map(Vec::from_iter)
+            .unwrap_or_default()
+            .into_iter()
```
This seems like unnecessary allocations, why not return the iterator that is returned from the `LruCache`?

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on protocols/request-response/src/lib.rs:
```diff
@@ -470,20 +470,14 @@ where
+    #[deprecated(note = "Use `libp2p_swarm::Swarm::add_peer_address` instead.")]
     pub fn add_address(&mut self, peer: &PeerId, address: Multiaddr) -> bool {
-        self.addresses.entry(*peer).or_default().insert(address)
+        self.addresses.put(*peer, std::iter::once(address))
```
This currently **overwrites** all addresses. I don't think that is what we want.

StemCll replied:
Changed the internals of `put`: https://github.com/libp2p/rust-libp2p/pull/4371/files#diff-cc3b2f08e48cc81069062816551b4be65ecce092128bca640d5d5024b6e9d6beR50

thomaseizinger commented on protocols/identify/src/behaviour.rs:
```diff
@@ -274,7 +273,19 @@ impl NetworkBehaviour for Behaviour {
+                            peer_id,
+                            addr: address.clone(),
+                        })
+                    });
```
I think this would be easier to read as a regular `for` loop. A regular loop would also avoid the allocation of the temporary vector for addresses to return (`self.events` may still have allocated capacity that we can just reuse).

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on Cargo.toml:
```diff
@@ -97,10 +97,10 @@ libp2p-pnet = { version = "0.24.0", path = "transports/pnet" }
-libp2p-swarm = { version = "0.44.1", path = "swarm" }
-libp2p-swarm-derive = { version = "=0.34.1", path = "swarm-derive" } # `libp2p-swarm-derive` may not be compatible with different `libp2p-swarm` non-breaking releases. E.g. `libp2p-swarm` might introduce a new enum variant `FromSwarm` (which is `#[non-exhaustive]`) in a non-breaking release. Older versions of `libp2p-swarm-derive` would not forward this enum variant within the `NetworkBehaviour`  ...
+libp2p-swarm-derive = { version = "=0.35.1", path = "swarm-derive" } # `libp2p-swarm-derive` may not be compatible with different `libp2p-swarm` non-breaking releases. E.g. `libp2p-swarm` might introduce a new enum variant `FromSwarm` (which is `#[non-exhaustive]`) in a non-breaking release. Older versions of `libp2p-swarm-derive` would not forward this enum variant within the `NetworkBehaviour`  ...
+libp2p-swarm = { version = "0.44.2", path = "swarm" }
```
Can we swap these two back so we avoid the diff?

thomaseizinger reviewed (commented):
Excellent work @StemCll !

Sorry for the delay in reviewing. I have a few final comments but overall, this is looking very good :)

Commit:
Revert version changes

Commit:
Extend swarm changelog entry; fix peer_addresses put logic

Commit:
Add additional test for remove

StemCll commented on Cargo.toml:
```diff
@@ -141,4 +141,8 @@ clippy.manual_let_else = "warn"
+    "/bin/sh",
+    '-c',
+    '/bin/sh $WORKSPACE_ROOT/scripts/add-changelog-header.sh',
+] # Nested use of shell to expand variables.
```
Formatter makes those changes automatically. I'll fix all of those unnecessary changes just before the merge

StemCll replied:
:heavy_check_mark:

StemCll commented on swarm/src/behaviour/peer_addresses.rs:
```diff
@@ -0,0 +1,306 @@
+                }
+            }
+
+            inserted_any
```
Wanted to do it with some iterator, but couldn't come up with a neat solution - thus the regular for loop :/

thomaseizinger replied:
Why do we need to take an `impl Iterator` as an argument? Accepting a single address is a more minimal API. Any caller can iterate themselves and then this function becomes much simpler. I'd also suggest we rename it to `add`.

Commit:
Add connection test to request response

Commit:
Rever toml file format changes

thomaseizinger commented on protocols/identify/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+## 0.44.2 - unreleased
+
+- Emit `ToSwarm::NewExternalAddrOfPeer` when identified new peer. `PeerCache` uses `libp2p_swarm::PeerAddresses`
+  See [PR 4371](https://github.com/libp2p/rust-libp2p/pull/4371)
```
```suggestion
- Emit `ToSwarm::NewExternalAddrOfPeer` for all external addresses of remote peers.
  See [PR 4371](https://github.com/libp2p/rust-libp2p/pull/4371).
```

thomaseizinger commented on protocols/autonat/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+
+- Allow temporary use of deprecated `libp2p_request_response::add_address`.
+  See [PR 4371](https://github.com/libp2p/rust-libp2p/pull/4371)
+
```
Let's remove this and not bump the version of `libp2p-autonat`. I'll apply the `internal-change` label to silence the CI check!

StemCll replied:
Alrighty :heavy_check_mark:

thomaseizinger commented on swarm/src/behaviour/peer_addresses.rs:
```diff
@@ -0,0 +1,306 @@
+                    let addr = prepare_addr(peer_id, addr);
+                    self.put(*peer_id, std::iter::once(addr));
+                    true
+                }
```
Couldn't we simply call our own `put` method here?

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on swarm/src/behaviour/peer_addresses.rs:
```diff
@@ -0,0 +1,306 @@
+                peer_id: Some(peer_id),
+                error: DialError::NoAddresses,
+                ..
+            }) => self.0.pop(peer_id).is_some(),
```
Sorry that I am just noticing this now but I don't think this makes any sense! `DialError::NoAddresses` means we didn't have any addresses to complete the dial. Purging the cache in this situation doesn't make any sense as we seem to not have any addresses anyway :)

Instead of `NoAddresses`, we should use the `Transport` variant and call `self.remove` for each address in there.

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on swarm/CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+- Add `libp2p::swarm::add_peer_address` that broadcasts address through `FromSwarm::NewExternalAddrOfPeer`.
+- Add `libp2p_swarm::behaviour::PeerAddresses` which is used as a cache for storing known addresses
+  of connected peers.
+  See [PR 4371](https://github.com/libp2p/rust-libp2p/pull/4371)
```
```suggestion
- Allow `NetworkBehaviour`s to share addresses of peers.
  This is enabled via the new `ToSwarm::NewExternalAddrOfPeer` event.
  The address is broadcast to all behaviours via `FromSwarm::NewExternalAddrOfPeer`.
  Protocols that want to collect these addresses can use the new `PeerAddresses` utility.
  See [PR 4371](https://github.com/libp2p/rust-libp2p/pull/4371).
```

thomaseizinger commented on swarm-derive/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+## 0.34.3 - unreleased
+
+- Generate code for `libp2p-swarm`'s `from_swarm::NewExternalAddrOfPeer` enum variant.
+  See [PR 4371](https://github.com/libp2p/rust-libp2p/pull/4371)
```
```suggestion
- Generate code for `libp2p-swarm`'s `FromSwarm::NewExternalAddrOfPeer` enum variant.
  See [PR 4371](https://github.com/libp2p/rust-libp2p/pull/4371).
```

thomaseizinger commented on swarm-derive/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+## 0.34.3 - unreleased
```
```suggestion
## 0.34.3
```

We no longer append `- unreleased` so that we can cut a release at any point.

thomaseizinger commented on swarm/CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+## 0.44.2 - unreleased
```
```suggestion
## 0.44.2
```

thomaseizinger commented on protocols/request-response/tests/peer_address.rs:
Awesome! Thank you for adding this test :)

thomaseizinger commented on protocols/request-response/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+## 0.26.2 - unreleased
```
```suggestion
## 0.26.2
```

thomaseizinger commented on protocols/request-response/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+## 0.26.2 - unreleased
+
+- Deprecate `libp2p_request_response::add_address`. `Behaviour` `addresses` field uses `libp2p_swarm::PeerAddresses`.
```
```suggestion
- Deprecate `Behaviour::add_address` in favor of `Swarm::add_peer_address`.
```

thomaseizinger commented on protocols/request-response/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+## 0.26.2 - unreleased
+
+- Deprecate `libp2p_request_response::add_address`. `Behaviour` `addresses` field uses `libp2p_swarm::PeerAddresses`.
+  See [PR 4371](https://github.com/libp2p/rust-libp2p/pull/4371)
```
```suggestion
  See [PR 4371](https://github.com/libp2p/rust-libp2p/pull/4371).
```

thomaseizinger commented on protocols/request-response/src/lib.rs:
```diff
@@ -470,20 +470,14 @@ where
     ///
     /// Returns true if the address was added, false otherwise (i.e. if the
     /// address is already in the list).
+    #[deprecated(note = "Use `libp2p_swarm::Swarm::add_peer_address` instead.")]
```
```suggestion
    #[deprecated(note = "Use `Swarm::add_peer_address` instead.")]
```

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on protocols/identify/src/behaviour.rs:
```diff
@@ -274,7 +273,20 @@ impl NetworkBehaviour for Behaviour {
+                        peer_id,
+                        addr: address.clone(),
+                    });
+                }
```
So instead of filtering here I'd suggest the following:

- As commented on `PeerAddresses`, change the API of `put` to only accept a single address.
- Call `self.discovered_peers.add` for each address
- If that returns `true`, append an event to `self.events`

This avoids the filtering altogether.

StemCll replied:
:heavy_check_mark: However maybe it would be good to merge this logic with what's happening in lines 271-272? Just to use `add` (instead of `put`) and check what was returned from `PeerAddresses`

thomaseizinger reviewed (commented):
I have a few more, mostly stylistic comments, otherwise this is ready to merge. Thank you for iterating on this with me :)

I am very excited for this to land. It will allow us to add the long-awaited [`PeerStore`](https://github.com/libp2p/rust-libp2p/issues/4103)!

thomaseizinger commented on swarm/src/lib.rs:
```diff
@@ -623,6 +624,17 @@ where
         self.confirmed_external_addr.remove(addr);
     }
 
+    /// Add peer's new external address.
```
```suggestion
    /// Add a new external address of a remote peer.
```

StemCll replied:
Thank you :pray:

thomaseizinger committed:
Documentation fixes

thomaseizinger commented on protocols/identify/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+## 0.44.2 - unreleased
```
```suggestion
## 0.44.2
```

thomaseizinger committed:
Update protocols/identify/CHANGELOG.md

Commit:
Revert autonat version bump

Commit:
Rename put to add in PeerAddresses

Commit:
Don't filter new external addresses

Commit:
Fix clippy warn

Commit:
Merge branch 'master' into feat/swarm/report_remote_address

thomaseizinger commented on protocols/identify/src/behaviour.rs:
```diff
@@ -445,42 +458,31 @@ fn multiaddr_matches_peer_id(addr: &Multiaddr, peer_id: &PeerId) -> bool {
     }
 
     fn put(&mut self, peer: PeerId, addresses: impl Iterator<Item = Multiaddr>) {
-        let cache = match self.0.as_mut() {
-            None => return,
-            Some(cache) => cache,
-        };
```
I'd prefer to stick to the original early-exit design.

1. It reduces the diff here.
2. It special-cases the non-happy path via early exit and keeps the happy path on the first level of indentation of the function.

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on protocols/request-response/src/lib.rs:
```diff
@@ -470,20 +470,14 @@ where
 
-    /// Removes an address of a peer previously added via `add_address`.
+    /// Removes an address of a peer previously added via [`Behaviour::add_address`].
     pub fn remove_address(&mut self, peer: &PeerId, address: &Multiaddr) {
```
We need to deprecate this one too. You can say in the message that it will be removed without replacement.

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on swarm/src/behaviour/peer_addresses.rs:
```diff
@@ -0,0 +1,329 @@
+
+fn prepare_addr(peer: &PeerId, addr: &Multiaddr) -> Multiaddr {
+    let addr = addr.clone();
+    addr.clone().with_p2p(*peer).unwrap_or(addr)
```
This isn't quite correct. `with_p2p` returns an error if the address has a **different** `/p2p` at the end. In that case, there is a mismatch between the provided `PeerId` and the address. That indicates a configuration error somewhere so I'd suggest that we don't add the address in that case (and return `false` from `add`).

StemCll replied:
:heavy_check_mark:

thomaseizinger commented on swarm/src/behaviour/peer_addresses.rs:
```diff
@@ -0,0 +1,329 @@
+                let address = prepare_addr(peer, address);
+                addrs.remove(&address)
+            },
+        )
```
I find this confusing to read. I'd suggest we do a `match` instead and use an early return of `return false` if we don't have entry for the given peer.

StemCll replied:
:heavy_check_mark:

thomaseizinger reviewed (commented):
Thanks! Sorry for the delay on this.

Let me know if you are able to work in a few more comments, otherwise I'll finalize this at some point in the next two weeks.

Commit:
Merge branch 'master' into feat/swarm/report_remote_address

Commit:
Review suggestions resolved

Commit:
Merge branch 'master' into feat/swarm/report_remote_address

thomaseizinger commented on protocols/identify/src/behaviour.rs:
```diff
@@ -274,7 +273,21 @@ impl NetworkBehaviour for Behaviour {
+                            });
+                        }
+                    }
+                }
```
Note for myself to add a test for this.

StemCll replied:
Thank you

thomaseizinger reviewed (dismissed):
Awesome! Thank you so much for this contribution and working in all my feedback :)

I've made a note for myself to add another test, will merge that in the upcoming days!

thomaseizinger committed:
Re-work `PeerAddresses` to be limited in how many addresses we store

This allows us to remove the `put` function and we can now safely
add all addresses individually to the cache. With the addition of
the `ToSwarm::NewExternalAddrOfPeer` event, the safeguard of
overwriting all addresses in the peer cache was no longer useful
because any other behaviour would have to apply this as well.

As such, we needed to build this safe-guard directly _into_ `PeerAddresses`.

This also makes the added test pass. Previously, the events would
never get emitted because after calling `.put`, the addresses were
already in the cache and the `add`-function always returned `false`.

thomaseizinger committed:
Allow new event in dctur test

thomaseizinger committed:
Fix clippy warnings

thomaseizinger committed:
Merge branch 'master' into feat/swarm/report_remote_address

mergify[bot] committed:
Merge branch 'master' into feat/swarm/report_remote_address
